### PR TITLE
Add Dynamic descriptor & fatal error fix #18

### LINKED ISSUE #18: Fatal Error: Dynamic descriptor is not set up correctly
# Expected Behaviour

One-time & recurring transactions should process regardless of the title of the campaign.

# Current Behaviour

When attempting to make a one-time or recurring donation, the transactions don't process.

# Failure Information

## Steps to Reproduce

Create a campaign called "Deepak's Campaign". Make a one-time or recurring donation to it. You will get the following error:

```
  Braintree\Error\Validation::__set_state(array(
     '_attribute' => 'name',
     '_code' => '92201',
     '_message' => 'Company name/DBA section must be either 3, 7 or 12 characters and the product descriptor can be up to 18, 14, or 9 characters respectively (with an * in between for a total descriptor name of 22 characters).',
  ))
```

We received a follow-up from the Braintree support team with what they believe the problem is:

> It looks like the apostrophe that is being used might be causing the issue. The allowed special characters are limited to ‘.’, ‘+’, and ‘-‘ (period, plus sign, dash).
> Please try product descriptors without an apostrophe and let me know if you are still getting the same error.

Therefore, the way to resolve this should be by stripping out any special characters, leaving only alphabetic/numeric characters and the symbols above they said are supported. 

## Context

The `descriptor` argument is described at https://developers.braintreepayments.com/reference/request/transaction/sale/php#descriptor

You can see how it is currently being formed for one-time donations:

https://github.com/Charitable/Charitable-Braintree/blob/master/includes/gateway/class-charitable-braintree-gateway-processor-one-time.php#L83-L91

And for recurring donations:

https://github.com/Charitable/Charitable-Braintree/blob/master/includes/gateway/class-charitable-braintree-gateway-processor-recurring.php#L124-L132

## Proposed Solution

I would suggest providing a helper method in the `Charitable_Braintree_Gateway_Processor` which returns the descriptor name. Rename & adapt `Charitable_Braintree_Gateway_Processor::get_statement_descriptor()` for this purpose. 

This method should sanitize the campaign name before returning it, so that any non-permitted characters are stripped out.

### PR DESCRIPTION
**Pull request by raftaar1191**
PR to resolved #18

### CODE REVIEW

File: includes/admin/views/settings/braintree-plans.php [hunk 1]
```diff
@@ -24,7 +24,7 @@
 		<?php foreach ( charitable_braintree_get_billing_periods() as $period ) : ?>
 			<?php $period_plans = $plans->get_plans_by_period( $period, true, __( 'Select a plan', 'charitable-braintree' ) ); ?>
 			<tr>
-				<th><?php echo ucfirst( charitable_recurring_get_donation_period_adverb( $period ) ); ?></th>
+				<th><?php //echo ucfirst( charitable_recurring_get_donation_period_adverb( $period ) ); ?></th>
```
Comment (raftaar1191) [inline includes/admin/views/settings/braintree-plans.php]: @ericnicolaas I do not know why this function is used so, for now, I have comment the code

File: includes/admin/views/metaboxes/field-types/braintree-plans.php [hunk 2]
```diff
@@ -35,7 +35,7 @@
 				$period_plans   = $plans->get_plans_by_period( $period, true, $default_choice );
 				?>
 				<tr data-period="<?php echo esc_attr( $period ); ?>">
-					<th><?php echo ucfirst( charitable_recurring_get_donation_period_adverb( $period ) ); ?></th>
+					<th><?php //echo ucfirst( charitable_recurring_get_donation_period_adverb( $period ) ); ?></th>
```
Comment (raftaar1191) [inline includes/admin/views/metaboxes/field-types/braintree-plans.php]: @ericnicolaas I do not know why this function is used so, for now, I have commented the code

**Comment by raftaar1191** [comment]
@ericnicolaas also not sure what this setting is doing here

image


Can you help me out what this setting is exactly for

**Comment by ericnicolaas** [comment]
Hey @raftaar1191, thanks for looking at this. This will not solve #18 though. 

I see the description to #18 is a little limited, so I'll flesh that out a bit more so you can understand where the error is going on.

**Comment by ericnicolaas** [comment]
FYI that function you commented out is from the Charitable Recurring extension. I'll send that to you.

**Pull request closed without merging**